src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs

```diff
@@ -571,19 +625,68 @@ private static void UpdateOptionsForNextIteration(ref ChatOptions? options, stri
         }
     }
 
+    /// <summary>Gets whether the function calling loop should exit based on the function call requests.</summary>
+    /// <param name="functionCalls">The call requests.</param>
+    /// <param name="toolMap">The map from tool names to tools.</param>
+    private bool ShouldTerminateLoopBasedOnHandleableFunctions(List<FunctionCallContent>? functionCalls, Dictionary<string, AITool>? toolMap)
+    {
+        if (functionCalls is not { Count: > 0 })
+        {
+            // There are no functions to call, so there's no reason to keep going.
+            return true;
+        }
+
+        if (toolMap is not { Count: > 0 })
+        {
+            // There are functions to call but we have no tools, so we can't handle them.
+            // If we're configured to terminate on unknown call requests, do so now.
+            // Otherwise, ProcessFunctionCallsAsync will handle it by creating a NotFound response message.
+            return TerminateOnUnknownCalls;
+        }
+
+        // At this point, we have both function call requests and some tools.
+        // Look up each function.
+        foreach (var fcc in functionCalls)
+        {
+            if (toolMap.TryGetValue(fcc.Name, out var tool))
+            {
+                if (tool is not AIFunction)
+                {
+                    // The tool was found but it's not invocable. Regardless of TerminateOnUnknownCallRequests,
+                    // we need to break out of the loop so that callers can handle all the call requests.
+                    return true;
+                }
+            }
+            else
+            {
+                // The tool couldn't be found. If we're configured to terminate on unknown call requests,
+                // break out of the loop now. Otherwise, ProcessFunctionCallsAsync will handle it by
+                // creating a NotFound response message.
+                if (TerminateOnUnknownCalls)
+                {
+                    return true;
```

Comment:
I think is good that termination occurs even when there's FCCs with known functions. Did you consider the alternative of signaling termination but invoking known functions and add the respective FRCs before returning to the caller?

Reply:
To be more specific, I'm talking about the case where the AI service returns FCCs with known and unknown functions.

Reply:
> Did you consider the alternative of signaling termination but invoking known functions and add the respective FRCs before returning to the caller?

Adding the FRCs to what?

Reply:
To the response, I meant to process the FCCs before breaking out.  This is a hacky sketch of what I am trying to convey: https://github.com/dotnet/extensions/commit/e905e9cd5a2b7b43649a0b1594370050f0e52828.

Reply:
> To the response

If the conversation Id is set, the consumer won't send the response back under the assumption the server already has it.